cp23/hpi-fun2/motivation/executables/generator.py

```diff
@@ -0,0 +1,221 @@
+    Path(filename + ".desc").write_text("Exponentially increasing weights of simple tree.")
+
+
+def single_node_bottleneck():
+    filename = "single_node_bottleneck"
+
+    test_case = (
+        "6 5\n"
+        "1 10 -8 -8 -8 -8\n"
+        "1 2 2 2 2"
+    )
+    Path(filename + ".in").write_text(test_case)
+    Path(filename + ".desc").write_text("There is a single bottleneck in the tree, which renders all its children unreachable.")
+
+
+def large_increasing_chain():
+    test_case = ""
+
+    employees = 100_000
+    threshold = 50_000
+    test_case += f"{employees} {threshold}\n"
+
+    levels = []
+    for i in range(0, employees):
+        levels.append(1)
+    test_case += " ".join([str(x) for x in levels]) + "\n"
+
+    bosses = []
+    for i in range(1, employees):
+        bosses.append(i)
+    test_case += " ".join([str(x) for x in bosses])
+
+    filename = "large_increasing_chain"
+    Path(filename + ".in").write_text(test_case)
+    Path(filename + ".desc").write_text("A long chain (only single child per node), where each child was weight 1.\nLimit Test.")
+
+
+def switching_chain():
+    test_case = ""
+
+    employees = 1_000
+    threshold = 1
+    test_case += f"{employees} {threshold}\n"
+
+    levels = []
+    for i in range(0, employees):
+        if i % 2 == 0:
+            levels.append(1)
+        else:
+            levels.append(-1)
+    test_case += " ".join([str(x) for x in levels]) + "\n"
+
+    bosses = []
+    for i in range(1, employees):
+        bosses.append(i)
+        test_case += f"{i} "
+    test_case += " ".join([str(x) for x in bosses]) + "\n"
+
+    filename = "switching_chain"
+    Path(filename + ".in").write_text(test_case)
+    Path(filename + ".desc").write_text("Long chain of nodes. Each node has either weight 1 or -1. No path length is larger 1.")
+
+
+def negative_only_test():
+    test_case = ""
+
+    hierarchical_levels = 8
+    employees = 0
+    for level in range(0, hierarchical_levels):
+        employees += 3**level
+
+    threshold = 20
+    test_case += f"{employees} {threshold}\n"
+
+    # Each person has negative motivational level
+    levels = []
+    for i in range(0, employees):
+        levels.append(-i)
+    test_case += " ".join([str(x) for x in levels]) + "\n"
+
+    # I did not think about it at first, but actually there is
+    #       (employees - 3**(hierarchical_levels-1))*3 = employees - 1
+    # Each person has exactly 3 people under them.
+    bosses = []
+    for i in range(1, employees - 3**(hierarchical_levels-1)):
+        bosses.extend([i, i, i])
+    test_case += " ".join([str(x) for x in bosses]) + "\n"
+
+    filename = "negative_only"
+    Path(filename + ".in").write_text(test_case)
+    Path(filename + ".desc").write_text("In this tree, each nodes has three children. All weights are negative, which render no child unreachable.")
+
+
+def random_test_case(case_number):
+    test_case = ""
+
+    employees = 100_000
```

Comment:
Please use global constants for the limits this ensures that they are more easily adjustable in future.

Reply:
**Important:** Our solutions have different outputs on the random samples. In general and for this purpose it would be nice to also have some smaller random testcases.

Reply:
I'm seeing a Seg Fault on your solution, cause by `large_incresing_chain.in`. I think it because if the large call stack caused by the chain of 100'000. It also might be only my machine, since that depends on local settings.

Reply:
But nevertheless I guess I found the problem, that was causing different results. I had a bug in my queue implementation.

BTW: Very lean implementation you did there! I did not think there was such a small solution, but very glad there is :D

Reply:
I'm thinking about the following: Even if we fix the seg fault in your implementation, it might occur in the contest and people will not have a lot of fun with a cause like this. So let's reduce this number and let the limiting tests get covered by the random test, which usually do not cause such immense call stacks.

Reply:
No, it won't occur ^^. Your OS gives every process a stack limit while your stack usage on the judge is unlimited (but within the bounds of the general very generous memory limit). If you are under UNIX you could try running `ulimit -s unlimited` before running the program. This should give the current terminal session and its subprocesses "unlimited" stack :)

Reply:
So keeping those tests should be fine :)

Reply:
TIL 💡

Reply:
Ok, I reduced the number of random test cases to 3 to have that.